src/commands/build-util.ts

```diff
@@ -1,19 +1,135 @@
+import {promises as fs} from 'fs';
 import type CSSModuleLoader from 'css-modules-loader-core';
-import type {EventEmitter} from 'events';
 import execa from 'execa';
+import {EventEmitter} from 'events';
 import npmRunPath from 'npm-run-path';
-import {
-  BuildScript,
-  SnowpackConfig,
-  SnowpackPluginBuildArgs,
-  SnowpackPluginBuildResult,
-} from '../config';
+import {BuildResult, SnowpackConfig, SnowpackSourceFile, SnowpackBuildMap} from '../config';
+import {getExt, replaceExt} from '../util';
 
 const IS_PREACT = /from\s+['"]preact['"]/;
 export function checkIsPreact(filePath: string, contents: string) {
   return filePath.endsWith('.jsx') && IS_PREACT.test(contents);
 }
 
+interface BuildFileOptions {
+  config: SnowpackConfig;
+  dest?: string;
+  messageBus?: EventEmitter;
+}
+
+/** Core Snowpack file pipeline builder */
+export async function buildFile(
+  srcPath: string,
+  destPath = srcPath,
+  options: BuildFileOptions,
+): Promise<SnowpackBuildMap> {
```

Comment:
Focus 1: create a “pure” build pipeline function that, given 1 input, generates 1 or more outputs. This `buildFile()` is the result of that: something that takes a source file, and runs it through the build pipeline. All that to say, this function is the heart of this PR.